engine/version/old_records_cleaner.cpp

```diff
@@ -8,6 +8,29 @@
 #include "../sorted_collection/skiplist.hpp"
 
 namespace KVDK_NAMESPACE {
+
+void OldRecordsCleaner::DelayFree(void* addr, TimeStampType ts) {
+  kvdk_assert(static_cast<DLRecord*>(addr)->IsDirty(), "");
+  kvdk_assert(access_thread.id >= 0, "");
+  auto& tc = cleaner_thread_cache_[access_thread.id];
+  std::lock_guard<SpinMutex> guard{tc.old_records_lock};
+  DataEntry* data_entry = static_cast<DataEntry*>(addr);
+  SpaceEntry entry{kv_engine_->pmem_allocator_->addr2offset_checked(addr),
+                   data_entry->header.record_size};
+  tc.pending_free_space_entries.push_back(PendingFreeSpaceEntry{entry, ts});
+
+  /// TODO: a thread may quit before all pending free entries are Free()d
+  /// GlobalClean() should collect those entries and Free() them.
+  maybeUpdateOldestSnapshot();
+  TimeStampType earliest_ts =
+      kv_engine_->version_controller_.OldestSnapshotTS();
+  while (tc.pending_free_space_entries.front().release_time < earliest_ts) {
```

Comment:
Set a limitation. You may free too many space entries here which can block foreground threads

Reply:
Ok. Maybe 64?